[Python 2 to 3] Bytes fixes

gpotter2 opened the pull request:
Here's the most boring part of it. That should bring neally all bytes support for the `core`. Contribution files will be changed later.

This PR:

- introduces some util functions in scapy.compat.py that works on both python 2 and 3:
-- raw(x) : returns the byte form of a string/packet/int/list of ints
-- plain_str(x) : does exactly the opposite: bytes/int ==> str
-- hex_bytes / bytes_hex ==> converts hex => bytes or bytes => hex
- Remove stupid BGP bug in a test: None ==> "None" ==> "\x04". Fixed by using `raw`

gpotter2 commented:
@p-l- @guedou Calling for review.

gpotter2 commented:
Ok I'm now able to boot scapy using python 3... Tons of stuff must be broken though...

gpotter2 commented:
Rebased.

mtury commented:
Looks fine as far as I can tell.

Please note that the TLS PR was merged last Thursday. I tried to follow your lead from PR #696 and such, but I'm not sure I caught every Python 2-exclusive statements. Hopefully you'll be able to tell quickly with whatever Python 3 platform you use for the convergence tests...?

gpotter2 commented:
Thank a lot for your work !
Pretty much everything was working ok.

Working to pass the tests...

gpotter2 commented:
Ok this is starting to look good...
I still have an unknown bug while importing the `mock` module :/

gpotter2 commented:
@guedou @p-l- Calling for review.

- Scapy will boot under python 3 with this code.
- I'm working on a next PR, based on this one, which will fix especially `regression.uts` format. Will push it once this one's done.

p-l- commented on scapy/contrib/bgp.uts:
```diff
@@ -297,7 +297,7 @@ str(BGPOptParam()) == b'\x02\x00'
 = BGPOptParam - Instantiation with specific values (1)
 str(BGPOptParam(param_type = 1)) == b'\x01\x00'
-str(BGPOptParam(param_type = 1, param_value = BGPAuthenticationInformation())) == b'\x01\x04'
+str(BGPOptParam(param_type = 1, param_value = BGPAuthenticationInformation())) == b'\x01\x00'
```
Why does this change?

gpotter2 replied:
Explanation: BGP was using i2m method that did not check if the field was empty. Here: "\x04" stands for `len(...) = 4` because the string "None" was passed to the function. It's a bug fixed indirectly

p-l- replied:
Great!

guedou commented:
See my commends in #738 Sorry for the inconvenience =/

p-l- commented on scapy/dadict.py:
```diff
@@ -11,13 +11,14 @@
 
 def fixname(x):
-    if x and x[0] in "0123456789":
+    if x and str(x[0]) in "0123456789":
```
If you need str(x[0]) here, you'll probably need str(x) the line after that I suppose.
Also, shouldn't we use `raw()` here instead of `str()`?

gpotter2 replied:
Below there's a use of `plain_str` to ensure every thing is stored as str. I think it's easier to load strings from files.
Can be discussed :/

p-l- commented on scapy/data.py:
```diff
@@ -78,19 +80,21 @@ def load_protocols(filename):
                     continue
                 dct[lt[0]] = int(lt[1])
             except Exception as e:
+                raise
```
Is that a debug code left?

gpotter2 replied:
Yes. Nice catch

p-l- commented on scapy/data.py:
```diff
@@ -78,19 +80,21 @@ def load_protocols(filename):
+                raise
                 log_loading.info("Couldn't parse file [%s]: line [%r] (%s)" % (filename,l,e))
     except IOError:
+        raise
```
Is that a debug code left?

p-l- commented on scapy/fields.py:
```diff
@@ -52,6 +53,8 @@ def i2count(self, pkt, x):
         return 1
     def h2i(self, pkt, x):
         """Convert human value to internal value"""
+        if type(x) is str:
```
Please use `isinstance(x, str)` instead.

p-l- commented on scapy/fields.py:
```diff
@@ -173,7 +176,7 @@ def getfield(self, pkt, s):
     def addfield(self, pkt, s, val):
         sval = self._fld.addfield(pkt, "", val)
-        return s+sval+struct.pack("%is" % (self.padlen(len(sval))), self._padwith)
+        return raw(s)+sval+struct.pack("%is" % (self.padlen(len(sval))), raw(self._padwith))
```
Are the `raw()` calls needed here? Be careful to call these functions only when necessary, the impact on performances could be important.

p-l- commented on scapy/fields.py:
```diff
@@ -209,7 +212,7 @@ def i2m(self, pkt, x):
         return str2mac(x)
     def any2i(self, pkt, x):
-        if isinstance(x, str) and len(x) is 6:
+        if isinstance(x, bytes) and len(x) is 6:
```
Can you use this opportunity to change that to `len(x) == 6`?

p-l- commented on scapy/fields.py:
```diff
@@ -300,7 +305,7 @@ class X3BytesField(XByteField):
         Field.__init__(self, name, default, "!I")
     def addfield(self, pkt, s, val):
-        return s+struct.pack(self.fmt, self.i2m(pkt,val))[1:4]
+        return raw(s)+struct.pack(self.fmt, self.i2m(pkt,val))[1:4]
```
Again, is `raw()` needed here?

p-l- commented on scapy/compat.py:
```diff
@@ -0,0 +1,140 @@
+                return bytes([x])
+            except ValueError:
+                # Ignore out of range strings
+                return bytes(str(x), "utf8")
```
OK this is dangerous!
`raw(255) == b'\xff'` while `raw(256) == b'256'`.

Also, do we need to have a specific case for `list` objects? They are handled by `bytes()` anyway.

Do we need to have an exception for `None`?

I think `raw()` should remain straightforward. Here is what we could have:
```Python
def raw(val):
    try:
        return bytes(val)
    except TypeError:
        return bytes(val, encoding="utf8")
```

And I think we could adapt the parts of the code that won't work with this. What do you think?

gpotter2 replied:
There got to be a special case for lists as using an encoding (
`bytes([1, 2, 3], encoding="utf8")`
) fails.

Otherwise this sounds fair.

p-l- commented on scapy/fields.py:
```diff
@@ -479,18 +486,18 @@ def getfield(self, pkt, s):
         return remain+ret,lst
     def addfield(self, pkt, s, val):
-        return s + "".join(str(v) for v in val)
+        return raw(s) + b"".join(raw(v) for v in val)
```
Same here: is `raw()` needed?

gpotter2 commented:
TODO:
- replace/(comment for others) as many `raw` and `plain_str` by `.encode(...)` and `.decode(...)` functions.

gpotter2 commented:
I'm splitting this PR up

gpotter2 commented:
Status update:
- Rebased
- Removed many `raw()` unneeded
- Some `raw()` are important: when `raw` bytes want to be converted from str to bytes, it's required to use `raw()` (non-ascii str)

guedou commented:
@gpotter2 don't forget to update the title when it its ready to be reviewed =)

gpotter2 commented:
Ready to be reviewed.

gpotter2 commented:
@guedou @p-l- Ready to be reviewed

gpotter2 commented:
I've added python 3 launchers

guedou commented on scapy/layers/inet6.py:
```diff
@@ -2630,7 +2636,7 @@ class MIP6OptAltCoA(_MIP6OptAlign, Packet):
     fields_desc = [ ByteEnumField('otype', 3, _mobopttypes),
                     ByteField('olen', 16),
-                    IP6Field("acoa", "::") ]
+                    IP6Field("acoa", b"::") ]
```
Is this really needed ?

gpotter2 replied:
No, nice catch

guedou commented on scapy/tools/UTscapy.py:
```diff
@@ -301,12 +302,18 @@ def dump_campaign(test_campaign):
 
-def sha1(x):
-    return sha.sha(x).hexdigest().upper()
+    def sha1(x):
```
Is the new indentation needed ?

gpotter2 replied:
Yes, the change log is a bit messy but the code is 
```
if six.PY2:
    def crc32(x):
        return "%08X" % (0xffffffff & zlib.crc32(x))

    def sha1(x):
         return hashlib.sha1(x).hexdigest().upper()
else:
    def crc32(x):
        return "%08X" % (0xffffffff & zlib.crc32(bytearray(x, "utf8")))

    def sha1(x):
        return hashlib.sha1(x.encode("utf8")).hexdigest().upper()
```

guedou commented:
It looks good to me !
